[Test][KubeRay]add doc test to kuberay-gcs-ft

popojk opened the pull request:
Implement doc tests described in [3157](https://github.com/ray-project/kuberay/issues/3157) for GCS fault tolerance in KubeRay.

## Related issue number

 [3157](https://github.com/ray-project/kuberay/issues/3157)

Doc URL: [https://anyscale-ray--51988.com.readthedocs.build/en/51988/cluster/kubernetes/user-guides/kuberay-gcs-ft.html](https://anyscale-ray--51988.com.readthedocs.build/en/51988/cluster/kubernetes/user-guides/kuberay-gcs-ft.html)
<!-- For example: "Closes #1234" -->

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

popojk commented:
hi @MortalHappiness , please help me to review this doctest PR. 

For your information, I ignore the test output of [this block](https://anyscale-ray--51988.com.readthedocs.build/en/51988/cluster/kubernetes/user-guides/kuberay-gcs-ft.html#step-5-create-a-detached-actor) and [this block](https://anyscale-ray--51988.com.readthedocs.build/en/51988/cluster/kubernetes/user-guides/kuberay-gcs-ft.html#step-8-access-the-detached-actor-again) because sometimes pod execution latency makes the count number not shown in log, making the doctest flaky. Please let me know if you have any suggestions, thanks!

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,760 @@
+    }
+   ],
+   "source": [
+    "# Step 4.1: List all Pods in the `default` namespace.\n",
+    "# The expected output should be 4 Pods: 1 head, 1 worker, 1 KubeRay, and 1 Redis.\n",
+    "kubectl get pods\n",
+    "\n",
+    "# Step 4.2: List all ConfigMaps in the `default` namespace.\n",
+    "kubectl get configmaps"
```
Split into 2 cells.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,760 @@
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 6.1: Check the RayCluster's UID.\n",
+    "kubectl get rayclusters.ray.io raycluster-external-redis -o=jsonpath='{.metadata.uid}'\n",
+    "# [Example output]: 864b004c-6305-42e3-ac46-adfa8eb6f752\n",
+    "\n",
+    "# Step 6.2: Check the head Pod's environment variable `RAY_external_storage_namespace`.\n",
+    "kubectl get pods $HEAD_POD -o=jsonpath='{.spec.containers[0].env}' | jq\n",
+    "# [Example output]:\n",
+    "# [\n",
+    "#   {\n",
+    "#     \"name\": \"RAY_external_storage_namespace\",\n",
+    "#     \"value\": \"864b004c-6305-42e3-ac46-adfa8eb6f752\"\n",
+    "#   },\n",
+    "#   ...\n",
+    "# ]\n",
+    "\n",
+    "# Step 6.3: Log into the Redis Pod.\n",
+    "# The password `5241590000000000` is defined in the `redis-config` ConfigMap.\n",
+    "# Step 6.4: Check the keys in Redis.\n",
+    "# Note: the schema changed in Ray 2.38.0. Previously we use a single HASH table,\n",
+    "# now we use multiple HASH tables with a common prefix.\n",
+    "export REDIS_POD=$(kubectl get pods --selector=app=redis -o custom-columns=POD:metadata.name --no-headers)\n",
+    "kubectl exec -i $REDIS_POD -- redis-cli -a \"5241590000000000\" KEYS '*'\n",
+    "# [Example output]:\n",
+    "# 1) \"RAY864b004c-6305-42e3-ac46-adfa8eb6f752@INTERNAL_CONFIG\"\n",
+    "# 2) \"RAY864b004c-6305-42e3-ac46-adfa8eb6f752@KV\"\n",
+    "# 3) \"RAY864b004c-6305-42e3-ac46-adfa8eb6f752@NODE\"\n",
+    "# [Example output Before Ray 2.38.0]:\n",
+    "# 2) \"864b004c-6305-42e3-ac46-adfa8eb6f752\"\n",
+    "#\n",
+    "\n",
+    "# Step 6.5: Check the value of the key.\n",
+    "kubectl exec -i $REDIS_POD -- redis-cli -a \"5241590000000000\" HGETALL RAY864b004c-6305-42e3-ac46-adfa8eb6f752@NODE\n",
+    "# Before Ray 2.38.0:\n",
+    "# HGETALL 864b004c-6305-42e3-ac46-adfa8eb6f752"
+   ]
+  },
```
Split into multiple cells.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,760 @@
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 7.1: Check the `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable in both the head Pod and worker Pod.\n",
+    "kubectl get pods $HEAD_POD -o=jsonpath='{.spec.containers[0].env}' | jq\n",
+    "# [Expected result]:\n",
+    "# No `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable is set. Hence, the Ray head uses its default value of `60`.\n",
+    "kubectl get pods $YOUR_WORKER_POD -o=jsonpath='{.spec.containers[0].env}' | jq\n",
+    "# [Expected result]:\n",
+    "# KubeRay injects the `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable with the value `600` to the worker Pod.\n",
+    "# [\n",
+    "#   {\n",
+    "#     \"name\": \"RAY_gcs_rpc_server_reconnect_timeout_s\",\n",
+    "#     \"value\": \"600\"\n",
+    "#   },\n",
+    "#   ...\n",
+    "# ]\n",
+    "\n",
+    "# Step 7.2: Kill the GCS process in the head Pod.\n",
+    "kubectl exec -i $HEAD_POD -- pkill gcs_server"
+   ]
```
ditto

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,760 @@
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 9.1: Delete the RayCluster custom resource.\n",
+    "kubectl delete raycluster raycluster-external-redis\n",
+    "\n",
+    "# Step 9.2: KubeRay operator deletes all Pods in the RayCluster.\n",
+    "# Step 9.3: KubeRay operator creates a Kubernetes Job to delete the Redis key after the head Pod is terminated.\n",
+    "\n",
+    "# Step 9.4: Check whether the RayCluster has been deleted.\n",
+    "kubectl get raycluster\n",
+    "# [Expected output]: No resources found in default namespace.\n",
+    "\n",
+    "# Step 9.5: Check Redis keys after the Kubernetes Job finishes.\n",
+    "export REDIS_POD=$(kubectl get pods --selector=app=redis -o custom-columns=POD:metadata.name --no-headers)\n",
+    "kubectl exec -i $REDIS_POD -- redis-cli -a \"5241590000000000\" KEYS \"*\"\n",
+    "# [Expected output]: (empty list or set)"
+   ]
```
ditto

MortalHappiness commented:
Hello. Please resolve conflicts. Thanks.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,900 @@
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": [
+     "nbval-ignore-output",
+     "remove-output"
+    ]
```
```suggestion
    "tags": [
     "nbval-ignore-output"
    ]
```
Ignore `nbval` for this cell but don't remove output for this cell. Also removes the `# [Example output]` above. ditto for step 6.2, 6.3, 6.4, 6.5, 7.1, 9.4, 9.5.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,900 @@
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 7.1: Check the `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable in both the head Pod and worker Pod.\n",
+    "kubectl get pods $HEAD_POD -o=jsonpath='{.spec.containers[0].env}' | jq\n",
+    "# [Expected result]:\n",
+    "# No `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable is set. Hence, the Ray head uses its default value of `60`.\n",
+    "kubectl get pods $YOUR_WORKER_POD -o=jsonpath='{.spec.containers[0].env}' | jq\n",
+    "# [Expected result]:\n",
+    "# KubeRay injects the `RAY_gcs_rpc_server_reconnect_timeout_s` environment variable with the value `600` to the worker Pod.\n",
+    "# [\n",
+    "#   {\n",
+    "#     \"name\": \"RAY_gcs_rpc_server_reconnect_timeout_s\",\n",
+    "#     \"value\": \"600\"\n",
+    "#   },\n",
+    "#   ...\n",
+    "# ]"
```
Split into 2 cells.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,900 @@
+   },
+   "outputs": [],
+   "source": [
+    "sleep 60\n",
+    "kubectl wait --for=condition=ready pod/$HEAD_POD --timeout=500s\n",
```
```suggestion
    "kubectl wait --for=condition=ready pod/$HEAD_POD --timeout=500s\n",
```

Don't need to sleep because we've already use `kubectl wait`.

MortalHappiness commented on doc/source/cluster/kubernetes/user-guides/kuberay-gcs-ft.ipynb:
```diff
@@ -0,0 +1,900 @@
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 9.1: Delete the RayCluster custom resource.\n",
+    "kubectl delete raycluster raycluster-external-redis"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "id": "0d00cf78-e7bd-403f-a0e9-e9c9fc9268e7",
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": [
+     "nbval-ignore-output",
+     "remove-output"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Step 9.2: KubeRay operator deletes all Pods in the RayCluster.\n",
+    "# Step 9.3: KubeRay operator creates a Kubernetes Job to delete the Redis key after the head Pod is terminated."
+   ]
+  },
```
Merge these 2 cells.

MortalHappiness commented:
@popojk I've left some review comments above. Thanks.

MortalHappiness commented:
Ping me again when this is ready for another review. Thanks.

popojk commented:
Hi @MortalHappiness , please review my updated PR. Let me mention a few points:
1. I removed the output of 6.5 and 9.5 as the outputs are empty.
2. In [this cell](https://github.com/ray-project/ray/pull/51988/files#diff-e1a3266afe6f74ff061f24be17a87edae73ecc023f33b1159d91368cf03e874dR769) I kept `sleep 60` because the ray-head pod doesn't restart immediately after executing `kubectl exec -i $HEAD_POD -- pkill gcs_server`. As a result, the `kubectl wait` command completes too early, causing the following doc tests to fail.

Thanks!

popojk commented:
Hi @MortalHappiness , just a friendly reminder that this PR has been updated based on your suggestions.
I've also addressed a few points in my previous reply. Please let me know if you have any further feedback. Thanks!

> Hi @MortalHappiness , please review my updated PR. Let me mention a few points:
> 
> 1. I removed the output of 6.5 and 9.5 as the outputs are empty.
> 2. In [this cell](https://github.com/ray-project/ray/pull/51988/files#diff-e1a3266afe6f74ff061f24be17a87edae73ecc023f33b1159d91368cf03e874dR769) I kept `sleep 60` because the ray-head pod doesn't restart immediately after executing `kubectl exec -i $HEAD_POD -- pkill gcs_server`. As a result, the `kubectl wait` command completes too early, causing the following doc tests to fail.
> 
> Thanks!

MortalHappiness commented:
LGTM. But as requested by @kevin85421, could you post a screenshot of running this test consecutively successful 10 times on localhost? Thanks.

kevin85421 commented:
cc @edoakes @jjyao would you mind merging this PR? Thanks!
